### DISCUSSION TITLE
Assigning to a variable before use in a conditional

### DISCUSSION THREAD

**Discussion by NoahCardoso** [JacquesCarette/Drasil#3869] [Q&A] [answered]
![image](https://github.com/user-attachments/assets/9e5745e7-89a7-48ad-bc85-17fbced8c22e)

I would like to be able to assign a variable before using it in the conditional. To do that I believe that I need to change `modules.hs` in `drasil-code` so that if the condition being checked uses a set that set is assigned to a variable before the condition is checked. I believe that the below function needs to be changed but I am not sure how to change it.

https://github.com/JacquesCarette/Drasil/blob/6033d408e7224438ad231bde91c411d440745909/code/drasil-code/lib/Language/Drasil/Code/Imperative/Modules.hs#L326-L334

**Comment by JacquesCarette** [comment]
@balacij or @B-rando1 should be able to help - as well as make sure your suggested change is the 'right' one to make. [I think what you're trying to do at a high level is right, but your suggested change doesn't seem quite right to me, as explained.]

**Comment by B-rando1** [comment, marked as answer]
I've looked a few times at this.  If we want to create the variable directly above the `if`-statement, this function is the place that makes the most sense to me.  Because it returns a list of statements, it would be difficult to insert the `varDecDef`s after this point.

I'm not exactly sure how it'll work with the types, but roughly what I think we need to do is modify the function like this:
```haskell
chooseConstr :: (OOProg r) => ConstraintBehaviour ->
  [(CodeVarChunk, [ConstraintCE])] -> GenState [MSStatement r]
chooseConstr cb cs = do
  varDefs <- mapM constrVarDef cs -- constrVarDef needs to act like `constrWarn` to generate either an `emptyStmt` or a `varDecDef`, depending on constraint type
  conds <- mapM (\(q,cns) -> mapM (convExpr . renderC q) cns) cs
  bods <- mapM (chooseCB cb) cs
  return $ concat $ zipWith $ (someting_here_probably (concat $ zipWith (zipWith (\cond bod -> ifNoElse [((?!) cond, bod)]))))
    varDefs conds bods
  where chooseCB Warning = constrWarn
        chooseCB Exception = constrExc
```

That's pretty rough and definitely wouldn't compile as-is 😅, but I _think_ that's roughly what you need to do.  Let me know if you need more to go on, I can look a bit more at the specifics if you need help.

**Reply by NoahCardoso** [reply]
Thank you! I will start working it. I will let you know if I need any further assistance.
